crates/rethnet_eth/src/remote.rs

```diff
@@ -97,72 +110,553 @@ impl BlockSpec {
     }
 }
 
-#[derive(serde::Serialize)]
-#[serde(untagged)]
-enum SerializableBlockSpec {
-    /// as a block number
-    #[serde(serialize_with = "serialize_u256")]
-    Number(U256),
-    /// as a block tag (eg "latest")
-    Tag(String),
+#[derive(Clone, Debug, PartialEq)]
+struct ZeroXPrefixedBytes {
+    inner: Bytes,
+}
+
+impl From<Bytes> for ZeroXPrefixedBytes {
+    fn from(b: Bytes) -> Self {
+        ZeroXPrefixedBytes { inner: b }
+    }
 }
 
-impl From<BlockSpec> for SerializableBlockSpec {
-    fn from(block_spec: BlockSpec) -> SerializableBlockSpec {
-        match block_spec {
-            BlockSpec::Number(n) => SerializableBlockSpec::Number(U256::from(n)),
-            BlockSpec::Tag(s) => SerializableBlockSpec::Tag(s),
+impl<'a> serde::Deserialize<'a> for ZeroXPrefixedBytes {
+    fn deserialize<D>(deserializer: D) -> Result<ZeroXPrefixedBytes, D::Error>
+    where
+        D: serde::Deserializer<'a>,
+    {
+        struct ZeroXPrefixedBytesVisitor;
+        impl<'a> serde::de::Visitor<'a> for ZeroXPrefixedBytesVisitor {
+            type Value = ZeroXPrefixedBytes;
+
+            fn expecting(&self, formatter: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+                formatter.write_str("a 0x-prefixed string of hex digits")
+            }
+
+            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                if &value[0..1] == "0x" {
```

Comment:
Thanks for noticing this. Yes, we do have tests of the serde logic in general (see all the tests i added in this PR), but we were not testing this particular branch. I added a test of this case in [this commit](https://github.com/NomicFoundation/hardhat/pull/3967/commits/f7953e759c1b1eb4d41f7475d15ca9d054eef2f4), and you can see the failure result for that commit in CI [here](https://github.com/NomicFoundation/hardhat/actions/runs/5180437673/jobs/9334716497#step:5:173), and you can see the "fix" [here](https://github.com/NomicFoundation/hardhat/pull/3967/commits/2cbf6f1aa2a0d77642332be347b954751b2b7b5e). Interestingly, note that the problem wasn't just the inverted boolean, but also the range operation: the way it was written before was exclusive of the ending range specifier; if it were *just* the inverted boolean, it would have been failing on all of the "working" test cases haha.

Reply:
Oh good catch, I missed that